enhancements/cloud-integration/infrastructure-external-platform-type.md

```diff
@@ -0,0 +1,586 @@
+and behaves the same as in the "None" platform case.
+In the future, we possibly might want to make CNO more tolerant of partner CNCC implementations
+and design a way for configuring platform-specific CNI behaviour.
+
+### Workflow Description
+
+During [phase 1](#implementation-phases) proposed changes intended to have no effect and the "External"
+platform type should be handled the same as the "None", so no specific user interaction is expected.
+
+### API Extensions
+
+A new optional constant of `PlatformType` type will be added to the "openshift/api".
+
+```go
+const (
+    ...
+    // ExternalPlatformType represent generic infrastructure provider. Provider-specific components should be supplemented separately.
+    ExternalPlatformType PlatformType = "External"
+    ...
+)
+```
+
+Additionally, the respective external platform spec and status should be added to the infrastructure resource.
+
+```go
+type CloudControllerManagerMode string
+
+const (
+    // Cloud Controller Manager is enabled and expected to be supplied.
+    // Signaling that kubelets and other CCM consumers should use --cloud-provider=external.
+    CloudControllerManagerExternal CloudControllerManagerMode = "External"
+    // Cloud Controller Manager is enabled and expected to be supplied.
+    // Signaling that kubelets and other CCM consumers should not set --cloud-provider flag.
+    CloudControllerManagerNone CloudControllerManagerMode = "None"
+)
+
+type CloudControllerManagerSettings struct {
+    // State determines whether or not an external Cloud Controller Manager is expected to 
+    // be presented in the cluster.
+    // For engaging an external Cloud Controller Manager, certain flags are expected to be set to the kubelets.
+    // https://kubernetes.io/docs/tasks/administer-cluster/running-cloud-controller/#running-cloud-controller-manager
+    //
+    // When set to "External", the respective operator (machine config operator) should set `--cloud-provider=external` flag to the kubelet.
+    // When omitted or disabled, no `cloud-provider` flag should be set.
+    // +kubebuilder:validation:Enum=External;None
+    // +optional
+    State CloudControllerManagerMode `json:"state,omitempty"`
+}
+
+
+// ExternalPlatformSpec holds the desired state for the generic External infrastructure provider.
+type ExternalPlatformSpec struct{
+    // ProviderName holds the arbitrary string represented cloud provider name, expected to be set at the installation time.
+    // Intended to serve only for informational purposes and not expected to be used for decision-making.
+    // +kubebuilder:default:="Unknown"
+    // +optional
+    ProviderName string `json:"providerName,omitempty"`
+    // CloudControllerManager contains settings specific to the external Cloud Controller Manager (a.k.a. CCM or CPI)
+    // +optional
+    CloudControllerManager CloudControllerManagerSettings `json:"cloudControllerManager"`
```

Comment:
Picking up where [this obsolete thread][1] left off, this config structure allows for:

* Both `--cloud-provider` and machine-API controllers:

    ```yaml
    platformSpec:
       type: External
       external:
         cloudControllerManager: External
    ```
* Machine-API controllers, but not `--cloud-provider`:

    ```yaml
    platformSpec:
       type: External
       external:
         cloudControllerManager: None
    ```
* Neither `--cloud-provider` nor machine-API controllers:

    ```yaml
    platformSpec:
       type: External
    ```

But once some other component grows `External`-specific logic (Cluster Cloud Controller Manager Operator?  Network operator?), is the plan to fill out `ExternalPlatformSpec` with knobs for each separate component?  Like:

```go
type ExternalPlatformSpec struct {
  ProviderName string `json:"providerName,omitempty"`
  CloudControllerManager CloudControllerManagerSettings `json:"cloudControllerManager"`
  MachineAPIManager MachineAPIManagerSettings `json:"machineAPIManager"`
  ComponentCManager ComponentCManagerSettings `json:"componentCManager"`
  ...
}
```

?

[1]: https://github.com/openshift/enhancements/pull/1234#discussion_r998800173

Reply:
Within this EP we supposed that MachineAPI related things would be enabled/disabled via the capabilities mechanism, at least for a while: https://github.com/openshift/enhancements/pull/1234/files#diff-08391a193550547988305aefa40e474c593e97d3b2375245c31015f39f5812b3R190

For now it seems to be suffice, and we was also asked to reduce code/api additions as much as possible.


>But once some other component grows External-specific logic (Cluster Cloud Controller Manager Operator? Network operator?), is the plan to fill out ExternalPlatformSpec with knobs for each separate component?

I do like this approach and would like to do so, but i can not be 100% certain now. Desired depth of the integration is a bit foggy for now, and we decided to limit this to a things which we will need for sure. Things like MachineAPI could be covered with capabilities, and there is no need to add it here right away.
There is a room to extend this api in this direction, and we certainly can go this route if we will need to.

Reply:
Being minimal makes sense. We know we need the cloud controller manager field. We don't yet know if we will ever need any others. Although there are examples of cases we could imagine them being useful, those are all potential future work that is not prioritized right now.